sql/src/main/java/io/crate/planner/Planner.java

```diff
@@ -377,5 +407,63 @@ private static LegacyUpsertById processInsertStatement(InsertFromValuesAnalyzedS
     public Functions functions() {
         return functions;
     }
+
+    private boolean allowSql(final AnalyzedStatement analyzedStatement) {
+        return hasValidLicense.getAsBoolean()
+               || analyzedStatement instanceof SetLicenseAnalyzedStatement
+               || analyzedStatement instanceof QueriedRelation;
+    }
+
+    private final class AllowQueryContext {
```

Comment:
I think there is no need to create a context to pass around a boolean.

It would probably be an option to decouple this altogether. Change `AllowQueryRelationVisitor` to something like `IsReadqueryOnSystemTable` (maybe implementing a `Predicate`) and than have a `ensureLicenseAllowsExecutionOfStatement` or something like that utilizing both.